### PR TITLE
Handle blocked cluster state in file settings

### PR DESCRIPTION
**Pull request by rjernst**
When file settings is first loaded on a master node starting up, the cluster state will likely be in a blocked state before it recovers. In that case the file settings will not be processable since the metadata will be missing in cluster state.

This commit makes watching for file settings not start until the cluster state is in a recovered state. It also updates the the reserved state update task to handle a similar case where a task may be queued and then run at time when the node is no longer master, but before the watcher is stopped.

### CODE REVIEW

File: server/src/test/java/org/elasticsearch/common/file/MasterNodeFileWatchingServiceTests.java [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        testService.clusterChanged(new ClusterChangedEvent("test", notRecoveredClusterState, ClusterState.EMPTY_STATE));
+        // just a master node isn't sufficient, cluster state also must be recovered
+        assertThat(testService.watching(), is(false));
+
+        ClusterState recoveredClusterState = ClusterState.builder(notRecoveredClusterState)
+            .blocks(ClusterBlocks.EMPTY_CLUSTER_BLOCK)
+            .build();
+        testService.clusterChanged(new ClusterChangedEvent("test", recoveredClusterState, notRecoveredClusterState));
+        // just a master node isn't sufficient, cluster state also must be recovered
+        assertThat(testService.watching(), is(true));
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/common/file/MasterNodeFileWatchingServiceTests.java:85]: Maybe also verify that we stop watching if either of those conditions becomes `false` again?

File: server/src/test/java/org/elasticsearch/reservedstate/service/ReservedStateUpdateTaskTests.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+import static org.hamcrest.Matchers.is;
+
+public class ReservedStateUpdateTaskTests extends ESTestCase {
+    public void testBlockedClusterState() {
+        var task = new ReservedStateUpdateTask("dummy", null, List.of(), Map.of(), List.of(), e -> {}, ActionListener.noop());
+        ClusterState notRecoveredClusterState = ClusterState.builder(ClusterName.DEFAULT)
+            .blocks(ClusterBlocks.builder().addGlobalBlock(GatewayService.STATE_NOT_RECOVERED_BLOCK))
+            .build();
+        assertThat(task.execute(notRecoveredClusterState), is(notRecoveredClusterState));
```
Comment (DaveCTurner) [inline server/src/test/java/org/elasticsearch/reservedstate/service/ReservedStateUpdateTaskTests.java:29]: nit: instance equality is what triggers the no-op detection in `MasterService`:

```suggestion
        assertThat(task.execute(notRecoveredClusterState), sameInstance(notRecoveredClusterState));
```

**Review by DaveCTurner — COMMENTED** [review]
Production changes LGTM, couple of tiny test comments

**Comment by rjernst** [comment]
Thanks @DaveCTurner. I've applied your suggestions.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Pull request merged**